aiosmtpd/smtp.py

```diff
@@ -461,17 +463,22 @@ async def smtp_HELP(self, arg):
         if arg:
             method = getattr(self, 'smtp_' + arg.upper(), None)
             if method and self._syntax_available(method):
-                help_str = method.__smtp_syntax__
-                if (self.session.extended_smtp
-                        and method.__smtp_syntax_extended__):
-                    help_str += method.__smtp_syntax_extended__
-                await self.push('250 Syntax: ' + help_str)
-                return
+                if (method != self.smtp_XCLIENT or
+                        self._is_handler_has_xclient_hook()):
+                    help_str = method.__smtp_syntax__
+                    if (self.session.extended_smtp
+                            and method.__smtp_syntax_extended__):
+                        help_str += method.__smtp_syntax_extended__
+                    await self.push('250 Syntax: ' + help_str)
+                    return
             code = 501
         commands = []
         for name in dir(self):
             if not name.startswith('smtp_'):
                 continue
+            if (name == 'smtp_XCLIENT') and (
+                    not self._is_handler_has_xclient_hook()):
+                continue
```

Comment:
With the new dict-based method lookup, this check should be done in `__init__` : if handler does not have XCLIENT hook, then remove `smtp_XCLIENT` method from the method dict.

See [line 174 of `smtp.py`](https://github.com/aio-libs/aiosmtpd/blob/16a89fe0054b3c04ccf204631f8fdc9cc85f95de/aiosmtpd/smtp.py#L174)

I think this change is also applicable, in a way for the [block starting line 466 of this commit](https://github.com/aio-libs/aiosmtpd/pull/138/files#diff-6dd0f8c95daec0ea7afa414a1a50c9f64a7ac4d95184662f8e211963e99349f0R466).